src/Compilers/CSharp/Portable/Symbols/ParameterSymbol.cs

```diff
@@ -385,6 +385,56 @@ internal sealed override ObsoleteAttributeData ObsoleteAttributeData
 
         internal abstract bool IsCallerMemberName { get; }
 
+        /// <summary>
+        /// If the method returns false, then we can infer that the argument used for this parameter was not null.
+        /// </summary>
+        internal abstract bool NotNullWhenFalse { get; }
+
+        /// <summary>
+        /// When the method returns, then we can infer that the argument used for this parameter was not null.
+        /// </summary>
+        internal abstract bool EnsuresNotNull { get; }
+
+        protected bool HasExtraAttribute(AttributeDescription description)
+        {
+            ParameterSymbol originalParameter = this.OriginalDefinition;
+            var containingMethod = originalParameter.ContainingSymbol as MethodSymbol;
+
+            if (containingMethod is null)
+            {
+                return false;
+            }
+
+            string key = ExtraAnnotations.MakeMethodKey(containingMethod);
```

Comment:
Should `ExtraAnnotations` be an instance, perhaps stored on `CSharpCompilationOptions`? That would allow specific testing in particular.

Reply:
Yes, that seems useful. I'll keep this for follow-up PR though.

---
In reply to: [187120948](https://github.com/dotnet/roslyn/pull/26656#discussion_r187120948) [](ancestors = 187120948)